### ISSUE TITLE
Headline A/B testing stats are not reliable

### ISSUE THREAD

**Issue by jace** [hasgeek/hasjob#216]
When a job headline is A/B tested, we do the following:

1. Any time the jobpost is impressed in an index, we assign an A/B group to the user's session (choice of A, B or neither).
2. When the user arrives via a social share, the group may be predetermined via the `b` URL parameter and is then assigned to their session.

* Group assignments are per-jobpost, so the user can be in `A` group for one and `B` group for another.
* A user can view alternate headlines by changing the value of the `b` parameter, but the initial group assignment persists for the duration of the session. Changing `b` does not count as a new impression for that group.
* Browsing sessions expire after half hour of inactivity or if the user clicks on a campaign link (anything with `utm_*` parameters). Browsing sessions are not related to login sessions, which are persistent.
* A user arriving in a new session may get assigned to a different group for the same jobpost and so may be shown the other headline this time.

In scenario 1, we assigned a grouping randomly and the user clicked through based on the choice we made. In scenario 2, no test was performed since the group was predetermined, but the visit still counts towards the success of either A or B headline.

**Why is this distinction important?** Imagine an influential tweeter sharing one of the headlines with their followers. Everyone who clicks through will count towards that headline's success, as they rightly should, but this does not automatically imply the other headline failed. The coin toss was done only once, for the first share. For an A/B test to be meaningful, the coin toss count is also important.

Hasjob should record both A/B group assignment and whether a coin toss was performed when assigning this group. Posters should get to see which headline performed better in a coin toss and which in subsequent shares.

**Comment by jace** [comment]
Since there are already separate `JobImpression` and `JobViewSession` models, both linking to the jobpost and user session, they can be presented as separate figures. However, JobImpression will have a roughly 50-50 split since it's *always* a coin toss. As such, this is not useful data. However, the overlap between these models will indicate which is working better.

Another possibility: both `JobImpression` and `JobViewSession` records exist, but with *different* A/B groups. This may happen if the job was impressed with one headline but the user ignored it, then followed a link from an external source with the other headline, within the same browsing session.

Grouping | NA | A | B | Explanation
-----|-----|-----|-----|-----
Impressions | NA | A | B | Number of times the job was listed
Conversions | NA | A | B | Responses to an A/B tested listing
External | NA | A | B | External links with a pre-specified A/B group
Crosslink | NA | A | B | Shown both and only responded to the second listing

**Comment by jace** [comment]
As a nice bonus of already having `JobImpression` and `JobViewSession` data separately, we have a full historical record and can introduce this feature without fresh data collection.

**Comment by jace** [comment]
We've implemented this and here are the [results from one listing](http://hasjob.hex.dev:5000/headstart.in/4b0mc):

Grouping | NA | A | B | Notes
-----|-----|-----|-----|-----
Impressions | 0 | 8891 | 11778 | The supposedly 50-50 coin toss seems to favour B. Should check the algorithm
Conversions | 10 | 430 | 555 | NA column should normally be 1, for the poster. The larger count is curious. Percentage figures would be nice here
External | 22 | 26 | 27 | Seems normal
Crosslink | 0 | 69 | 5 | NA should be 0 as expected since a crosslink is only when B-group is non-null

**Comment by jace** [comment]
Added percentage-based conversion display in 4e4b049.

**Comment by jace** [comment]
Changed from `bool(randint(0, 1))` to `random.choice([True, False])` in e706f0e. Hopefully this will eliminate the bias towards 1.

**Comment by jace** [comment]
`random.choice` appears to be working as the figures are now within 5% of each other. However, e752e3c introduced a serious bug.

When a `JobViewSession` is recorded, we look for a pre-existing `JobImpression` to set the `cointoss` flag. However, `JobViewSession` is created within the request cycle while `JobImpression` is created in a background job that may be running several seconds late. As a result, they may be created out of order, recording a conversion as an external link.

In 6832a758ce00091a9b82c2360bd690307bf75fce the creation of `JobViewSession` has also been deferred to a background job that hopefully does not run out of order (since there are two queue runners in production and the two jobs may be in parallel when there is a light load).

**Comment by jace** [comment]
`random.choice([True, False])` still shows a bias towards `True`, ie the first item in the list, so we're now using this (hat tip @karthikb351): `bool(datetime.utcnow().microsecond % 2)`

**Comment by jace** [comment]
Back to `random.choice` in 5a21a2e2effeabd77d42fe429236570df0e025fb using [John von Neumann's procedure](https://en.wikipedia.org/wiki/Fair_coin#Fair_results_from_a_biased_coin):

```python
def cointoss():
    r1 = r2 = None
    while r1 == r2:
        r1 = choice([True, False])
        r2 = choice([True, False])
    return r1
```

**Comment by karthikb351** [comment]
Well, this makes no sense. I find it hard to believe the algorithm has a bias. In any case, I ran some tests against the four algorithms we considered.

#### randint
```python
def rand_int():
    return bool(randint(0, 1))
```

### random.choice
```python
def random_choice():
    return random.choice([True, False])
```

### microsecond%2
```python
def microsecond_mod():
    return bool(datetime.utcnow().microsecond % 2)
```

### faircoin
```python
def faircoin():
    r1 = r2 = None
    while r1 == r2:
        r1 = random.choice([True, False])
        r2 = random.choice([True, False])
    return r1
```

I samples against all four algorithms on the production server, with 1k, 10k, 100k coin tosses. Geek stats below

### Results
Method | 1k tosses | 10k tosses | 100k tosses | Time for single toss
-----|-----|-----|-----|-----|-----
randint | 499 H, 501 T,  Error of ~0.1% | 4955 H, 5045 T,  Error of ~0.45% | 50104 H, 49896 T,  Error of ~0.104% | 2.35969ms
random.choice | 491 H, 509 T,  Error of ~0.9% | 4976 H, 5024 T,  Error of ~0.24% | 49744 H, 50256 T,  Error of ~0.256% | 1.83402ms
microsecond%2 | 493 H, 507 T,  Error of ~0.7% | 5004 H, 4996 T,  Error of ~0.04% | 49989 H, 50011 T,  Error of ~0.011% | 3.60124ms
faircoin | 499 H, 501 T,  Error of ~0.1% | 5072 H, 4928 T,  Error of ~0.72% | 49974 H, 50026 T,  Error of ~0.026% | 3.8942ms

**Comment by jace** [comment]
`microsecond%2` and `faircoin` having such different results based on the number of times you toss them makes no sense. Seems… random.

**Comment by jace** [comment]
Turns out the bias bug was elsewhere. Fixed in bd6a86a.
